File: src/core/auditor.ts
import { AuditReport, SkillContext, BaseCheck } from './types';
import { DangerousToolsCheck } from './checks/dangerous-tools';
import * as crypto from 'crypto';

export class Auditor {
  private checks: BaseCheck[] = [new DangerousToolsCheck()];

  async audit(context: SkillContext): Promise<AuditReport> {
    const results = await Promise.all(this.checks.map(c => c.run(context)));
    
    const weightedScore = results.reduce((acc, r) => {
      const check = this.checks.find(c => c.id === r.id);
      return acc + (r.score * (check?.weight || 1));
    }, 0);
    
    const totalWeight = this.checks.reduce((acc, c) => acc + c.weight, 0);

    return {
      skillHash: crypto.createHash('sha256').update(context.raw).digest('hex'),
      finalScore: totalWeight > 0 ? weightedScore / totalWeight : 10,
      results,
      timestamp: new Date().toISOString()
    };
  }
}

File: internal/engine/auditor.go
package engine

import (
	"context"
	"crypto/sha256"
	"fmt"
	"sync"
	"time"
	"github.com/albertowar/skillauditai/pkg/api"
	"github.com/albertowar/skillauditai/internal/checks"
	"github.com/albertowar/skillauditai/internal/behavioral"
)

type Auditor struct {
	Checks     []checks.Check
	Behavioral *behavioral.Service
}

func NewAuditor(b *behavioral.Service) *Auditor {
	return &Auditor{
		Checks:     checks.AllChecks(),
		Behavioral: b,
	}
}

func (a *Auditor) Audit(ctx context.Context, skill api.SkillContext) (api.AuditReport, error) {
	var wg sync.WaitGroup
	results := make([]api.CheckResult, len(a.Checks))
	
	for i, c := range a.Checks {
		wg.Add(1)
		go func(idx int, check checks.Check) {
			defer wg.Done()
			res, err := check.Run(ctx, skill, a.Behavioral)
			if err != nil {
				results[idx] = api.CheckResult{
					ID:            check.ID(),
					Name:          check.Name(),
					Score:         0,
					Level:         api.Critical,
					Justification: fmt.Sprintf("Check failed: %v", err),
				}
			} else {
				results[idx] = res
			}
		}(i, c)
	}

	wg.Wait()

	var weightedScore float64
	var totalWeight float64
	for i, r := range results {
		w := a.Checks[i].Weight()
		weightedScore += r.Score * w
		totalWeight += w
	}

	finalScore := 10.0
	if totalWeight > 0 {
		finalScore = weightedScore / totalWeight
	}

	hash := sha256.Sum256([]byte(skill.Raw))

	return api.AuditReport{
		SkillHash:  fmt.Sprintf("%x", hash),
		FinalScore: finalScore,
		Results:    results,
		Timestamp:  time.Now(),
	}, nil
}
